safe_control_gym/controllers/mpc/mpc_acados.py

```diff
@@ -0,0 +1,391 @@
+        acados_model.u_labels = self.env.ACTION_LABELS
+        acados_model.t_label = 'time'
+
+        self.acados_model = acados_model
+
+    @timing
+    def compute_initial_guess(self, init_state, goal_states=None):
+        '''Use IPOPT to get an initial guess of the solution.
+
+        Args:
+            init_state (ndarray): Initial state.
+            goal_states (ndarray): Goal states.
+        '''
+        x_val, u_val = super().compute_initial_guess(init_state, goal_states)
+        self.x_guess = x_val
+        self.u_guess = u_val
+
+    def setup_acados_optimizer(self):
+        '''Sets up nonlinear optimization problem.'''
+        nx, nu = self.model.nx, self.model.nu
+        ny = nx + nu
+        ny_e = nx
+
+        # create ocp object to formulate the OCP
+        ocp = AcadosOcp()
+        ocp.model = self.acados_model
+
+        # set dimensions
+        ocp.dims.N = self.T  # prediction horizon
+
+        # set cost (NOTE: safe-control-gym uses quadratic cost)
+        ocp.cost.cost_type = 'LINEAR_LS'
+        ocp.cost.cost_type_e = 'LINEAR_LS'
+        ocp.cost.W = scipy.linalg.block_diag(self.Q, self.R)
+        ocp.cost.W_e = self.Q if not self.use_lqr_gain_and_terminal_cost else self.P
+        ocp.cost.Vx = np.zeros((ny, nx))
+        ocp.cost.Vx[:nx, :nx] = np.eye(nx)
+        ocp.cost.Vu = np.zeros((ny, nu))
+        ocp.cost.Vu[nx:(nx + nu), :nu] = np.eye(nu)
+        ocp.cost.Vx_e = np.eye(nx)
+        # placeholder y_ref and y_ref_e (will be set in select_action)
+        ocp.cost.yref = np.zeros((ny, ))
+        ocp.cost.yref_e = np.zeros((ny_e, ))
+
+        # Constraints
+        # general constraint expressions
+        state_constraint_expr_list = []
+        input_constraint_expr_list = []
+        for sc_i, state_constraint in enumerate(self.state_constraints_sym):
+            state_constraint_expr_list.append(state_constraint(ocp.model.x))
+        for ic_i, input_constraint in enumerate(self.input_constraints_sym):
+            input_constraint_expr_list.append(input_constraint(ocp.model.u))
+
+        h_expr_list = state_constraint_expr_list + input_constraint_expr_list
+        h_expr = cs.vertcat(*h_expr_list)
+        h0_expr = cs.vertcat(*h_expr_list)
+        he_expr = cs.vertcat(*state_constraint_expr_list)  # terminal constraints are only state constraints
+        # pass the constraints to the ocp object
+        ocp = self.processing_acados_constraints_expression(ocp, h0_expr, h_expr, he_expr)
+
+        # slack costs for nonlinear constraints
+        if self.soft_constraints:
+            # slack variables for all constraints
+            ocp.constraints.Jsh_0 = np.eye(h0_expr.shape[0])
+            ocp.constraints.Jsh = np.eye(h_expr.shape[0])
+            ocp.constraints.Jsh_e = np.eye(he_expr.shape[0])
+            # slack penalty
+            L2_pen = self.soft_penalty
+            L1_pen = self.soft_penalty
+            ocp.cost.Zl_0 = L2_pen * np.ones(h0_expr.shape[0])
```

Comment:
should we soften the initial constraint?

Reply:
And are we softening the input constraints? we shouldnt but not sure if we are

Reply:
I agree with the initial condition part. For the input constraint part I mainly want to keep the consistency with the current MPC formulation. @adamhall any suggestion on this?

Reply:
I agree with the initial condition part. For the input constraints I mainly keep the consistency with the current MPC formulation. @adamhall any suggestion on this?

Reply:
I checked the acados formulation [(eq (2) and (3))](https://github.com/acados/acados/blob/master/docs/problem_formulation/problem_formulation_ocp_mex.pdf It seems that the initial condition is a separate constraint and would be affected by the softening. The softening will affect the other constraints acting on the initial node and I think this should not affect the optimization much.